tests/unit/plugins/test_publisher.py

```diff
@@ -15,22 +15,409 @@
+
     success, error = asyncio.Queue(), asyncio.Queue()
-    client = publisher.GDNSPublisher(config, success, error)
+    plugin = publisher.GDNSPublisher(config, success, error, client)
 
-    assert interfaces.IPublisherClient.providedBy(client)
+    assert interfaces.IPublisherClient.providedBy(plugin)
     assert interfaces.IPublisherClient.implementedBy(publisher.GDNSPublisher)
-    assert config is client.config
-    assert success is client.success_channel
-    assert error is client.error_channel
-    assert 'publish' == client.phase
+    assert config is plugin.config
+    assert success is plugin.success_channel
+    assert error is plugin.error_channel
+    assert 'publish' == plugin.phase
+
+
+@pytest.mark.asyncio
+async def test_publish_changes_raises_exception_on_invalid_zone(
+        publisher_instance, event_message, event_msg_data_with_invalid_zone,
+        caplog):
+    """Ensure exception raised on invalid zone"""
+    event_message.data = event_msg_data_with_invalid_zone
+
+    with pytest.raises(exceptions.InvalidDNSZoneInMessageError) as error:
+        await publisher_instance.publish_changes(
+            event_message)
+    assert error.match('Error when asserting zone for record:')
+    assert 2 == len(caplog.records)
+
+
+@pytest.mark.asyncio
+async def test_publish_changes_handles_update_conflict(
+    publisher_instance, event_message,
+        all_existing_zone_records, rrsets_url, changes_url,
+        initial_changes_req, handled_conflict_changes_req,
+        initial_changes_pending_json_resp, initial_changes_resp, caplog):
+    """Ensure exception raised on invalid zone"""
+    expected_change_id = json.loads(initial_changes_pending_json_resp)['id']
+    event_message.data['resourceRecords'] = initial_changes_req['additions']
+    publisher_instance.http_client._request_post_mock.side_effect = [
+        exceptions.GCPHTTPError('409'), initial_changes_pending_json_resp]
+    publisher_instance.http_client._get_all_mock.return_value = (
+        all_existing_zone_records)
+    publisher_instance.http_client._get_json_mock.side_effect = [
+        initial_changes_resp]
+
+    await publisher_instance.publish_changes(event_message)
+
+    publisher_instance.http_client._get_all_mock.assert_called_with(rrsets_url)
+    publisher_instance.http_client._request_post_mock.assert_called_with(
+        'post', changes_url, json=handled_conflict_changes_req)
+    publisher_instance.http_client._get_json_mock.assert_called_with(
+        f'{changes_url}/{expected_change_id}')
+    assert 3 == len(caplog.records)
+
+
+@pytest.mark.asyncio
+async def test_publish_changes_raises_exception_on_publish_timeout(
+    publisher_instance, event_message, changes_url,
+        initial_changes_req, initial_changes_resp,
+        initial_changes_pending_json_resp, mock_sleep, caplog):
+    """Ensure exception raised when publish wait timeout exceeded."""
+    publisher_instance.publish_wait_timeout = 0.0001
```

Comment:
Did we want to try the mocking datetime approach?

Reply:
i'm OK with leaving this as is.